File: sdk/python/lib/pulumi/next/resource.py
# Copyright 2016-2018, Pulumi Corporation.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""The Resource module, containing all resource-related definitions."""
from typing import Optional, List, Any

from ..runtime import known_types
from ..runtime.resource import register_resource, register_resource_outputs
from ..runtime.settings import get_root_resource
from ..runtime.unknown import Unknown


class ResourceOptions:
    """
    ResourceOptions is a bag of optional settings that control a resource's behavior.
    """

    parent: Optional['Resource']
    depends_on: Optional[List['Resource']]
    protect: Optional[bool]

    def __init__(self,
                 parent: Optional['Resource'] = None,
                 depends_on: Optional[List['Resource']] = None,
                 protect: Optional[bool] = None) -> None:
        self.parent = parent
        self.depends_on = depends_on
        self.protect = protect


class Resource:
    """
    Resource represents a class whose CRUD operations are implemented by a provider plugin.
    """

    """
    The stable, logical URN used to distinctly address a resource, both before and after deployments.
    """
    urn: str

    """
    The provider-assigned unique ID for this managed resource.  It is set during deployments and may
    be missing during planning phases.
    """
    id: str

    def __init__(self,
                 t: str,
                 name: str,
                 custom: bool,
                 props: Optional[dict] = None,
                 opts: Optional[ResourceOptions] = None) -> None:
        if not t:
            raise TypeError('Missing resource type argument')
        if not isinstance(t, str):
            raise TypeError('Expected resource type to be a string')
        if not name:
            raise TypeError('Missing resource name argument (for URN creation)')
        if not isinstance(name, str):
            raise TypeError('Expected resource name to be a string')

        # Properties and options can be missing; simply, initialize to empty dictionaries.
        if props:
            if not isinstance(props, dict):
                raise TypeError('Expected resource properties to be a dictionary')
        elif not props:
            props = dict()
        if opts:
            if not isinstance(opts, ResourceOptions):
                raise TypeError('Expected resource options to be a ResourceOptions instance')
        if not opts:
            opts = ResourceOptions()

        # Default the parent if there is none.
        if opts.parent is None:
            opts.parent = get_root_resource()

        # Now register the resource.  If we are actually performing a deployment, this resource's properties
        # will be resolved to real values.  If we are only doing a dry-run preview, on the other hand, they will
        # resolve to special Preview sentinel values to indicate the value isn't yet available.
        result = register_resource(t, name, custom, props, opts)

        # Set the URN, ID, and output properties.
        self.urn = result.urn
        if result.id:
            self.id = result.id
        else:
            self.id = Unknown()

        if result.outputs:
            self.set_outputs(result.outputs)

    def set_outputs(self, outputs: dict):
        """
        Sets output properties after a registration has completed.
        """


@known_types.custom_resource
class CustomResource(Resource):
    """
    CustomResource is a resource whose create, read, update, and delete (CRUD) operations are managed
    by performing external operations on some physical entity.  The engine understands how to diff
    and perform partial updates of them, and these CRUD operations are implemented in a dynamically
    loaded plugin for the defining package.
    """

    """
    id is the provider-assigned unique ID for this managed resource.  It is set during
    deployments and may be missing (undefined) during planning phases.
    """
    id: str

    """
    CustomResource is a resource whose CRUD operations are managed by performing external operations on some
    physical entity.  Pulumi understands how to diff and perform partial updates ot them, and these CRUD operations
    are implemented in a dynamically loaded plugin for the defining package.
    """
    def __init__(self,
                 t: str,
                 name: str,
                 props: Optional[dict] = None,
                 opts: Optional[ResourceOptions] = None) -> None:
        Resource.__init__(self, t, name, True, props, opts)


class ComponentResource(Resource):
    """
    ComponentResource is a resource that aggregates one or more other child resources into a higher level
    abstraction.  The component itself is a resource, but does not require custom CRUD operations for provisioning.
    """
    def __init__(self,
                 t: str,
                 name: str,
                 props: Optional[dict] = None,
                 opts: Optional[ResourceOptions] = None) -> None:
        Resource.__init__(self, t, name, False, props, opts)
        self.id = None

    def register_outputs(self, outputs):
        """
        Register synthetic outputs that a component has initialized, usually by allocating other child
        sub-resources and propagating their resulting property values.
        """
        if outputs:
            register_resource_outputs(self, outputs)


def output(name: str, value: Any):
    """
    Exports a named stack output.
    """
    stack = get_root_resource()
    if stack is not None:
        stack.output(name, value)

File: sdk/go/common/apitype/migrate/resource.go
// Copyright 2016-2018, Pulumi Corporation.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package migrate

import (
	"github.com/pulumi/pulumi/sdk/v3/go/common/apitype"
	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
)

// UpToResourceV2 migrates a resource from ResourceV1 to ResourceV2.
func UpToResourceV2(v1 apitype.ResourceV1) apitype.ResourceV2 {
	var v2 apitype.ResourceV2
	v2.URN = v1.URN
	v2.Custom = v1.Custom
	v2.Delete = v1.Delete
	v2.ID = v1.ID
	v2.Type = v1.Type
	v2.Inputs = make(map[string]interface{})
	for key, value := range v1.Inputs {
		v2.Inputs[key] = value
	}
	// v1.Defaults was deprecated in v2.
	v2.Outputs = make(map[string]interface{})
	for key, value := range v1.Outputs {
		v2.Outputs[key] = value
	}
	v2.Parent = v1.Parent
	v2.Protect = v1.Protect
	// v2.External is a new field that, when true, indicates that this resource's
	// lifecycle is not owned by Pulumi. Since all V1 resources have their lifecycles
	// owned by Pulumi, this is `false` for all V1 resources.
	v2.External = false
	// v2.Provider is a reference to a first-class provider associated with this resource.
	v2.Provider = ""
	v2.Dependencies = append(v2.Dependencies, v1.Dependencies...)
	v2.InitErrors = append(v2.InitErrors, v1.InitErrors...)
	return v2
}

// UpToResourceV3 migrates a resource from ResourceV2 to ResourceV3.
func UpToResourceV3(v2 apitype.ResourceV2) apitype.ResourceV3 {
	var v3 apitype.ResourceV3
	v3.URN = v2.URN
	v3.Custom = v2.Custom
	v3.Delete = v2.Delete
	v3.ID = v2.ID
	v3.Type = v2.Type
	v3.Inputs = v2.Inputs
	v3.Outputs = v2.Outputs
	v3.Parent = v2.Parent
	v3.Protect = v2.Protect
	v3.External = v2.External
	v3.Dependencies = v2.Dependencies
	v3.InitErrors = v2.InitErrors
	v3.Provider = v2.Provider

	// v3.PropertyDependencies tracks dependencies on a per-input-property basis. We conservatively assume that all
	// properties depend on all of the resource's dependencies.
	propertyDependencies := make(map[resource.PropertyKey][]resource.URN)
	for pk := range v3.Inputs {
		propertyDependencies[resource.PropertyKey(pk)] = v3.Dependencies
	}
	v3.PropertyDependencies = propertyDependencies

	return v3
}
